### DISCUSSION TITLE
Issue with AudioPlayer when using a mix of http and https urls

### DISCUSSION THREAD

**Discussion by cschwick** [pschatzmann/arduino-audio-tools#1446] [Q&A] [answered]
Dear Phil,

First a big thank you for this spectacular masterpiece of software... impressive !

I think I found a minor issue but not being sure if I misunderstood things, I first post it here in the discussions.

I modified slightly one of the examples (player-url-i2s) using a mixture of https and http urls.

I think I found a feature in URLStream.h which prevents using a mixture of secure and insecure urls (http and https): 

Line 355: f (client != nullptr) return *client;

If the client variable has previously been set in the Object, it will be reused. But this means, if you switch from http to https you will always use the same client (the one which was created the first time an URL was used in the object). Hence, when you started with an http url an "insecure" client is used to handle a https request later. This times out and does not work. If you put the https URL as the first URL in the list, it is presented as first url to the URLStream object and it works. However, if you then want to play a http url it fails since the "WIFIClintSecure" cannot handle insecure urls.

I solved this by brute force commenting out the line mentioned above. I think this does not do any damage since at maximum one WiFiClient and one WiFiCLientSecure will be constructed and then we are good for all urls. I also checked that if the url is changed (in URLStream.preProcess(...) afterwards the getClient(...) method is called.

Not sure if I missed anything... I do not have a good overview of this software yet and in addition I do not work with the Arduino ecosystem (so far I used the bare bone idf system of Espressif).

The problem can be re-produced with the modifed example below (just changed the url list and some pins for my setup).

I hope this message is useful and I did not oversee something ... 

Thanks again for this super library!

Chris

------

I use the following enviroment:

PlatformIO with arduino framework. This is my plaformio.ini:

--------------
~~~
[env:esp32dev]
platform = espressif32 @ 5.4.0
board = esp32dev
framework = arduino
lib_deps =
   Wire
   https://github.com/pschatzmann/arduino-audio-tools.git
   https://github.com/pschatzmann/arduino-libhelix.git
build_flags =  -Wno-unused-variable -Wno-unused-but-set-variable -Wno-unused-function -Wno-format-extra-args -O2 -DCOPY_LOG_OFF
monitor_speed = 115200
board_build.f_cpu = 240000000L
~~~
-------------

audio-tools library 0.9.6

And here my slightly modified example code:

----------
~~~
/**
 * @file player-url-i2s.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-url-i2s/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */


#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"


const char *urls[] = {
  "http://stream.srg-ssr.ch/m/rsj/mp3_128",
  "http://stream.srg-ssr.ch/m/drs3/mp3_128",
  "http://stream.srg-ssr.ch/m/rr/mp3_128",
  "https://playerservices.streamtheworld.com/api/livestream-redirect/TLPSTR20.mp3",
  "http://streaming.swisstxt.ch/m/drsvirus/mp3_128",
  "http://22593.live.streamtheworld.com/SOULRADIO.mp3",
  "http://sc6.radiocaroline.net:8040/mp3"
};
const char *wifi = "****";
const char *password = "****";

URLStream urlStream(wifi, password);
AudioSourceURL source(urlStream, urls, "audio/mp3");
I2SStream i2s;
MP3DecoderHelix decoder;
AudioPlayer player(source, i2s, decoder);

// additional controls
const int volumePin = A0;
Debouncer nextButtonDebouncer(2000);
const int nextButtonPin = 2;

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT;
  cfg.port_no = 1;
  cfg.pin_ws = 22;
  cfg.pin_bck = 21;
  cfg.pin_data = 4; 
  i2s.begin(cfg);

  // setup player
  player.begin();
}

// Sets the volume control from a linear potentiometer input
void updateVolume() {
  // Reading potentiometer value (range is 0 - 4095)
  float vol = static_cast<float>(analogRead(volumePin));
  // min in 0 - max is 1.0
  player.setVolume(vol/4095.0);
}


// Moves to the next url when we touch the pin
void updatePosition() {
   if (digitalRead(nextButtonPin)) {
      Serial.println("Moving to next url");
      if (nextButtonDebouncer.debounce()){
        player.next();
      }
  }
}


void loop() {
  //updateVolume(); // remove comments to activate volume control
  updatePosition();  // remove comments to activate position control
  player.copy();
}
~~~

-------------------- 

The modified "getClient" routine in URLStream (Just inserted commented out the first if-statement)
~~~
  /// Determines the client
  Client& getClient(bool isSecure) {
    // if (client != nullptr) return *client;  allow mixture of http and https urls !!!
#ifdef USE_WIFI_CLIENT_SECURE
    if (isSecure) {
      if (clientSecure == nullptr) {
        clientSecure = new WiFiClientSecure();
        clientSecure->setInsecure();
      }
      LOGI("WiFiClientSecure");
      return *clientSecure;
    }
#endif
#ifdef USE_WIFI
    if (clientInsecure == nullptr) {
      clientInsecure = new WiFiClient();
      LOGI("WiFiClient");
    }
    return *clientInsecure;
#else
    LOGE("Client not set");
    stop();
    return *client;  // to avoid compiler warning
#endif
  }
~~~

**Comment by pschatzmann** [comment, marked as answer]
You are right: I did not expect to process mixed http and https requests since I had mainly [this usage example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-url_mp3_helix-i2s/streams-url_mp3_helix-i2s.ino) in mind. 
I committed a [correction](https://github.com/pschatzmann/arduino-audio-tools/commit/c5334858a3f5b4e30302511a6006cee059399ea0) which  keeps the different client implementations better apart.

**Comment by cschwick** [comment]
Great! 
1000 Dank/Merci !
